pkg/sources/github_experimental/object_discovery.go

```diff
@@ -0,0 +1,654 @@
+          %s
+        }
+      }
+    `, owner, repo, testCases)
+
+	headers := map[string]string{
+		"Authorization":         "Bearer " + ghToken,
+		"Content-Type":          "application/json",
+		"Github-Verified-Fetch": "true",
+		"X-Requested-With":      "XMLHttpRequest",
+		"Accept-Language":       "en-US,en;q=0.9",
+		"Priority":              "u=1, i",
+	}
+
+	requestBody, err := json.Marshal(map[string]string{"query": query})
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request body: %w", err)
+	}
+
+	req, err := http.NewRequest("POST", "https://api.github.com/graphql", bytes.NewBuffer(requestBody))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+
+	for key, value := range headers {
+		req.Header.Set(key, value)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("python request error: %w", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	var data responseData
+	if err := json.Unmarshal(body, &data); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
+	}
+
+	if len(data.Errors) > 0 {
+		return nil, fmt.Errorf("%s (GitHub Request Error)", strings.Split(data.Errors[0].Message, ".")[0])
+	}
+	if data.Message != "" {
+		return nil, fmt.Errorf("%s (GitHub Request Error)", strings.Split(data.Message, ".")[0])
+	}
+
+	commits := data.Data.Repository
+
+	valid_cfor := []string{}
+	invalid := []string{}
+
+	for commit, value := range commits {
+		commit = strings.Replace(commit, "commit", "", 1)
+		if value.OID == "{}" || value.OID == "" {
+			invalid = append(invalid, commit)
+		} else {
+			valid_cfor = append(valid_cfor, value.OID)
+		}
+	}
+
+	res := map[string][]string{
+		validHiddenCommit: valid_cfor,
+		invalidCommit:     invalid,
+	}
+
+	return res, nil
+}
+
+// createBatches divides a slice into batches of a specified size
+func createBatches(items []string, batchSize int) <-chan []string {
+	out := make(chan []string)
+	go func() {
+		defer close(out)
+		itemsCopy := append([]string(nil), items...)
+		for len(itemsCopy) > 0 {
+			end := batchSize
+			if len(itemsCopy) < batchSize {
+				end = len(itemsCopy)
+			}
+			batch := itemsCopy[:end]
+			itemsCopy = itemsCopy[end:]
+			out <- batch
+		}
+	}()
+	return out
+}
+
+// downloadPatches fetches and checks out cfor commits
+func downloadPatches(valid_cfor []string, path string) error {
+	// Download all patches
+	for batch := range createBatches(valid_cfor, gitFetchMax) {
+		gitArgs := []string{
+			"-C",
+			path,
+			"--work-tree",
+			path,
+			"fetch",
+			"--quiet",
+			"origin",
+		}
+		gitArgs = append(gitArgs, batch...)
+		_, err := runGitCommand(gitArgs)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Checkout each commit
+	// Note: path and worktree are needed or else git will do something funny with the actual cwd
```

Comment:
@joeleonjr What is the purpose of checking out each commit after fetching it?

(Also, what "funny" thing does Git do? :o)

Reply:
Not an expert, but this is my best understanding: When we call `git fetch origin <commit hash>`, it's downloading those commit objects to your local. However, it's not attached to any branch or working state until you check it out. As a result, if you run `trufflehog` against your local without first checking out that commit, it won't get scanned. It's possible a config update to the way we run git scanning would address this and we could remove the need to check out. If you want to test it out:

```
git clone https://github.com/joeleonjr/cfor_test
cd cfor_test 
trufflehog git file://.
git fetch origin 92398a91fde2f43813723a0ed52e64e6967f7e8e
trufflehog git file://.         
git checkout 92398a91fde2f43813723a0ed52e64e6967f7e8e
trufflehog git file://.           
```

Only the last time you run `trufflehog` above, will you see a canary token.